modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -0,0 +1,284 @@
+        Metadata old = currentDatabases.get(name);
+        currentDatabases.put(name, new Metadata(System.currentTimeMillis(), old.getFirstChunk(), old.getLastChunk(), old.getMd5()));
+        logger.info("geoip database [" + name + "] is up to date, updated timestamp");
+        updateTaskState(name);
+    }
+
+    boolean updateTaskState(String name) {
+        PlainActionFuture<PersistentTask<?>> future = PlainActionFuture.newFuture();
+        persistentTask.updatePersistentTaskState(state, future);
+        try {
+            state = ((GeoIpTaskState) future.actionGet().getState()).copy();
+            return true;
+        } catch (Exception e) {
+            logger.error("can not update state during geoip database [" + name + "] update", e);
+            return false;
+        }
+    }
+
+    //visible for testing
+    int indexChunks(String name, InputStream is, int chunk) throws IOException {
+        for (byte[] buf = getChunk(is); buf.length != 0; buf = getChunk(is)) {
+            client.prepareIndex(DATABASES_INDEX).setId(name + "_" + chunk)
+                .setSource(XContentType.SMILE, "name", name, "chunk", chunk, "data", buf)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setWaitForActiveShards(ActiveShardCount.ALL)
+                .get();
+            chunk++;
+        }
+        return chunk;
+    }
+
+    //visible for testing
+    byte[] getChunk(InputStream is) throws IOException {
+        byte[] buf = new byte[MAX_CHUNK_SIZE];
+        int chunkSize = 0;
+        while (chunkSize < MAX_CHUNK_SIZE) {
+            int read = is.read(buf, chunkSize, MAX_CHUNK_SIZE - chunkSize);
+            if (read == -1) {
+                break;
+            }
+            chunkSize += read;
+        }
+        if (chunkSize < MAX_CHUNK_SIZE) {
+            buf = Arrays.copyOf(buf, chunkSize);
+        }
+        return buf;
+    }
+
+    //for testing only
+    void setPersistentTask(AllocatedPersistentTask persistentTask) {
+        this.persistentTask = persistentTask;
+    }
+
+    //for testing only
+    void setState(GeoIpTaskState state) {
+        this.state = state;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask allocatedTask, PersistentTaskParams params, PersistentTaskState state) {
+        if (scheduled != null) {
+            scheduled.cancel();
+        }
+        if (enabled) {
+            this.persistentTask = allocatedTask;
+            this.state = state != null ? (GeoIpTaskState) state : new GeoIpTaskState();
+            try {
+                updateDatabases();
+            } catch (Exception e) {
+                logger.error("exception during geoip databases update", e);
+            }
+            scheduleNextRun(pollInterval);
+        }
+    }
+
+    private void scheduleNextRun(TimeValue time) {
+        scheduled = threadPool.schedule(() -> nodeOperation(persistentTask, null, state), time, ThreadPool.Names.GENERIC);
```

Comment:
I think we should add some randomization to the time value here. Otherwise every cluster is going to send a request at the top of the hour. Would we great if can spread it over an entire hour. (for example add `randomBetween(0, 60*60*1000)`)

Reply:
I think it will be scheduled `time` amount of millis after last run. So depending on when task was started request time will differ. I think on global scale this should be enough randomization, WDYT?

Reply:
True, that should be enough randomized.